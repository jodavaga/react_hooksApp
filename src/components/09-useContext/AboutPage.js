import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const AboutPage = () => {

    const { user, setUser } = useContext( UserContext );

    const { username } = user;

    return (
        <div className="container">
            <h1>AboutPage</h1>
            <hr />

            <h4>Hola!, <strong>{ username ? username : 'Por favor haga login' }</strong>...</h4>
        </div>
    )
}
