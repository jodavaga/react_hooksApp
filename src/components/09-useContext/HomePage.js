import React, { useContext } from 'react';
import { UserContext } from './UserContext';
import { useHistory } from 'react-router-dom';

export const HomePage = () => {

    const { user, setUser } = useContext( UserContext );

    const history = useHistory();

    const handleLogout = () => {
        setUser({});

        alert('Loged out!');
        history.push('/login');
    }

    return (
        <div className="container">
            <h1>HomePage</h1>
            <hr />
            <div className="d-flex justify-content-end">
                <button
                    className="btn btn-outline-danger"
                    onClick={ handleLogout }
                >
                    Logout
                </button>
            </div>

            <pre>
                { JSON.stringify(user, null, 3) }
            </pre>
        </div>
    )
}
