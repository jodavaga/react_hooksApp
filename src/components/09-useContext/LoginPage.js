import React, { useContext } from 'react';
import { useForm } from '../../hooks/useForm';
import { UserContext } from './UserContext';
import { useHistory } from 'react-router-dom';

export const LoginPage = () => {

    const [ {username, password}, handleInputChange ]  = useForm({
        username: '',
        password: ''
    });

    // useContext (use of UserContext)
    const { user, setUser } = useContext( UserContext );

    // router navigation
    const history = useHistory();
    

    const handleSubmit = (e) => {
        e.preventDefault();

        // setValues on context
        setUser({
            ...user,
            username,
            password
        });

        // navigate to HomePage after submit
        // history.push('/');

    }

    return (
        <div className="container">
            <h1>LoginPage</h1>
            <hr />
            
            <form onSubmit={ handleSubmit }>
                <div className="form-group">
                    <input
                        className="form-control"
                        type="text"
                        placeholder="username"
                        name='username'
                        value={ username }
                        onChange={ handleInputChange }
                        autoComplete="off"
                    />
                </div>

                <div className="form-group">
                    <input
                        className="form-control"
                        type="password"
                        placeholder="password"
                        name='password'
                        value={ password }
                        onChange={ handleInputChange }
                        autoComplete="off"
                    />
                </div>

                <button
                    className="btn btn-primary mt-3"
                    type="submit"
                >
                    Guardar
                </button>
            </form>

            <br/>
            <pre>   
                {/* get values from context */}
                { user.username && `usuario registrado: ${ JSON.stringify(user, null, 3) }` }
            </pre>



        </div>
    )
}
