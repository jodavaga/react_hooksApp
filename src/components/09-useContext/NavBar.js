import React from 'react';

import { Link, NavLink } from 'react-router-dom';

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            <a className="navbar-brand" href="/">useContext</a>
            <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavAltMarkup" aria-controls="navbarNavAltMarkup" aria-expanded="false" aria-label="Toggle navigation">
                <span className="navbar-toggler-icon"></span>
            </button>
            <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                <NavLink exact activeClassName="active"  to="/" className="nav-item nav-link">Home</NavLink>
                <NavLink exact activeClassName="active"  to="/about" className="nav-item nav-link">About</NavLink>
                <NavLink exact activeClassName="active"  to="/login" className="nav-item nav-link">Login</NavLink>
                </div>
            </div>
        </nav>
    )
}
