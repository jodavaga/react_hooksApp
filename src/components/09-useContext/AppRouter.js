import React from 'react';
import { 
    BrowserRouter as Router,
    Switch,
    Route,
    Link
 } from 'react-router-dom';

import { LoginPage } from './LoginPage';
import { AboutPage } from './AboutPage';
import { HomePage } from './HomePage';
 
export const AppRouter = () => {
    return (
        <Router>
            <div>

            </div>

            <Switch>
                <Route exact path="/" component={ HomePage }></Route>
                <Route exact path="/login" component={ LoginPage }></Route>
                <Route exact path="/about" component={ AboutPage }></Route>
            </Switch>
        </Router>
    )
}
