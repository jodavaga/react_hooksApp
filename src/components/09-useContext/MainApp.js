import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

    // const user = {
    //     id: 1,
    //     name: "Jose D. Vasquez",
    //     email: "email@email.com"
    // }

    const [user, setUser] = useState({})

    return (
        <UserContext.Provider value={{
            user, 
            setUser
        }}>
            <AppRouter />

        </UserContext.Provider>

    )
}
