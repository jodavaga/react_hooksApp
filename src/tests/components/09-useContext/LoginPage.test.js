import React from 'react';
import { mount } from 'enzyme';

import { LoginPage } from '../../../components/09-useContext/LoginPage';
import { UserContext } from '../../../components/09-useContext/UserContext';


describe('Test on <LoginPage />', () => {

    const user = {
        username: 'Jose D.',
        password: '12345'
    }
    const setUser = jest.fn();
    // using mount will render all child components included in <HomePage />
    const wrapper = mount(
        <UserContext.Provider value={{
            user,
            setUser
        }}>
            <LoginPage />
        </UserContext.Provider>
    );

    test('should render component', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('should setUser on submit', () => {
        

        const formSubmit = wrapper.find('form').prop('onSubmit');
        // simulate submit
        formSubmit({ preventDefault(){} });


        expect(setUser).toHaveBeenCalled();
    });
    
    
})
