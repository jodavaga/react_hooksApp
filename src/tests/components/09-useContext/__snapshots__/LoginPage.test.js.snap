// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <LoginPage /> should render component 1`] = `
<div
  className="container"
>
  <h1>
    LoginPage
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoComplete="off"
        className="form-control"
        name="username"
        onChange={[Function]}
        placeholder="username"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        autoComplete="off"
        className="form-control"
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary mt-3"
      type="submit"
    >
      Guardar
    </button>
  </form>
  <br />
  <pre>
    usuario registrado: {
   "username": "Jose D.",
   "password": "12345"
}
  </pre>
</div>
`;
